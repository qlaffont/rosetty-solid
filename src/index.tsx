/* eslint-disable @typescript-eslint/no-explicit-any */
import { Language, rosetty, RosettyReturn } from 'rosetty';
import {
  Accessor,
  createContext,
  createMemo,
  createSignal,
  JSX,
  useContext,
} from 'solid-js';

export const RosettyContext = createContext();

export const RosettyProvider = (props: {
  children: JSX.Element;
  languages: Record<string, Language>;
  defaultLanguage: string;
  translateFallback?: boolean;
}) => {
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  //@ts-ignore
  const r = createMemo(
    () =>
      rosetty(props.languages, props.defaultLanguage, props.translateFallback),
    []
  );

  const [actualLang, setActualLang] = createSignal(props.defaultLanguage);

  const changeLang = (lang: string) => {
    r().changeLang(lang);
    setActualLang(r().getCurrentLang()!);
  };

  return (
    <RosettyContext.Provider
      value={{
        ...r(),
        actualLang,
        changeLang,
      }}
    >
      {props.children}
    </RosettyContext.Provider>
  );
};

type AnyObject = Record<string, any>;

export function useRosetty<T extends AnyObject>(): RosettyReturn<T> & {
  actualLang: Accessor<string | undefined>;
} {
  const client = useContext(RosettyContext);

  if (!client) {
    throw new Error('No RosettyClient set, use RosettyProvider to set one');
  }

  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  //@ts-ignore
  return client;
}

export type Rosetty<T extends AnyObject> = RosettyReturn<T>;
