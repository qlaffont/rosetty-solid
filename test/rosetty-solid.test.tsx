/**
 * @jest-environment jsdom
 */
/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable @typescript-eslint/ban-ts-comment */
// Importing the jest testing library
import '@testing-library/jest-dom';

import { cleanup, renderHook } from '@solidjs/testing-library';
//@ts-ignore
import enGB from 'dayjs/locale/en-gb';
//@ts-ignore
import fr from 'dayjs/locale/fr';

import { RosettyContext, RosettyProvider, useRosetty } from '../src';
//@ts-ignore
const I18NContextProvider = (props) => (
  <RosettyProvider
    languages={props.languages}
    defaultLanguage={props.defaultLanguage}
  >
    {props.children}
  </RosettyProvider>
);

afterEach(() => {
  cleanup();
});

describe('rosetty solid', () => {
  it('should export useI18n + HOC + context', () => {
    expect(typeof useRosetty).toBe('function');
    expect(typeof RosettyProvider).toBe('function');
    expect(typeof RosettyContext).toBe('object');
  });

  it('should be able to use i18n', () => {
    //@ts-ignore
    const wrapper = (props) => (
      <I18NContextProvider
        languages={{ fr: { dict: {}, locale: fr } }}
        defaultLanguage="fr"
      >
        {props.children}
      </I18NContextProvider>
    );
    const { result } = renderHook(() => useRosetty(), { wrapper });
    expect(result.languages).toStrictEqual(['fr']);
    expect(Object.keys(result)).toStrictEqual([
      'changeLang',
      'languages',
      'getCurrentLang',
      't',
      'displayNames',
      'listFormat',
      'numberFormat',
      'pluralRules',
      'format',
      'formatRelative',
      'formatDistance',
      'formatDistanceToNow',
      'formatDuration',
      'actualLang',
    ]);
  });

  it('should be able to return error', () => {
    //@ts-ignore
    const wrongWrapperLanguageNotValid = (props) => (
      <I18NContextProvider
        languages={{ fr: { dict: {}, locale: fr } }}
        defaultLanguage="en"
      >
        {props.children}
      </I18NContextProvider>
    );

    try {
      renderHook(() => useRosetty(), {
        wrapper: wrongWrapperLanguageNotValid,
      });
    } catch (error) {
      //@ts-ignore
      expect(error.message).toBe('rosetty: language en not found');
    }
  });

  it('should be able to return actualLang', () => {
    //@ts-ignore
    const wrapper = (props) => (
      <I18NContextProvider
        languages={{
          fr: { dict: { toto: 'test' }, locale: fr },
          en: { dict: { toto: 'EN' }, locale: enGB },
        }}
        defaultLanguage="en"
      >
        {props.children}
      </I18NContextProvider>
    );

    const { result, cleanup } = renderHook(() => useRosetty(), { wrapper });

    expect(result.actualLang()).toStrictEqual('en');
    //@ts-ignore
    expect(result.t('toto')).toStrictEqual('EN');
    result.changeLang('fr');
    cleanup();
    expect(result.actualLang()).toStrictEqual('fr');
    //@ts-ignore
    expect(result.t('toto')).toStrictEqual('test');
  });
});
