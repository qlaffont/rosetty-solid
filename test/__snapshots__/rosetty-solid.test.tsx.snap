// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rosetty solid should be able to use i18n 1`] = `
[
  "changeLang",
  "languages",
  "getCurrentLang",
  "t",
  "displayNames",
  "listFormat",
  "numberFormat",
  "pluralRules",
  "dateTimeFormat",
  "relativeTimeFormat",
  "collator",
  "segmenter",
  "durationFormat",
  "actualLang",
]
`;
